packages/rpc-methods/src/restricted/dialog.ts

```diff
@@ -0,0 +1,298 @@
+ * @returns The specification for the `snap_dialog` permission.
+ */
+const specificationBuilder: PermissionSpecificationBuilder<
+  PermissionType.RestrictedMethod,
+  DialogSpecificationBuilderOptions,
+  DialogSpecification
+> = ({
+  allowedCaveats = null,
+  methodHooks,
+}: DialogSpecificationBuilderOptions) => {
+  return {
+    permissionType: PermissionType.RestrictedMethod,
+    targetKey: methodName,
+    allowedCaveats,
+    methodImplementation: getDialogImplementation(methodHooks),
+  };
+};
+
+export const dialogBuilder = Object.freeze({
+  targetKey: methodName,
+  specificationBuilder,
+  methodHooks: {
+    showAlert: true,
+    showConfirmation: true,
+    showPrompt: true,
+  },
+} as const);
+
+type AlertParameters = {
+  type: DialogType.alert;
+  fields: AlertFields;
+};
+
+type ConfirmationParameters = {
+  type: DialogType.confirmation;
+  fields: ConfirmationFields;
+};
+
+type PromptParameters = {
+  type: DialogType.prompt;
+  fields: PromptFields;
+};
+
+export type DialogParameters =
+  | AlertParameters
+  | ConfirmationParameters
+  | PromptParameters;
+
+/**
+ * Builds the method implementation for `snap_dialog`.
+ *
+ * @param hooks - The RPC method hooks.
+ * @param hooks.showAlert - A function that shows an alert in the MetaMask UI
+ * and returns when the user has closed the alert.
+ * @param hooks.showConfirmation - A function that shows a dialog in the
+ * MetaMask UI and returns a `boolean` that signals whether the user
+ * approved or denied the confirmation.
+ * @param hooks.showPrompt - A function that shows a prompt in the MetaMask UI
+ * and returns the value the user entered into the prompt's text field.
+ * @returns The method implementation which returns `true` if the user approved the confirmation, otherwise `false`.
+ */
+export function getDialogImplementation({
+  showAlert,
+  showConfirmation,
+  showPrompt,
+}: DialogMethodHooks) {
+  // This rule appears to trigger because ESLint does not understand execution
+  // will never reach the "end" of this function.
+  // eslint-disable-next-line consistent-return
+  return async function dialogImplementation(
+    args: RestrictedMethodOptions<DialogParameters>,
+  ): Promise<boolean | null | string> {
+    const {
+      params,
+      context: { origin },
+    } = args;
+
+    const { type, fields } = getValidatedParams(params);
+    switch (type) {
+      case DialogType.alert:
+        return showAlert(origin, fields);
+
+      case DialogType.confirmation:
+        return showConfirmation(origin, fields);
+
+      case DialogType.prompt:
+        return showPrompt(origin, fields);
+
+      /* istanbul ignore next */
+      default:
+        assertExhaustive(type);
+    }
+  };
+}
+
+// TODO(rekmarks): Use an OpenRPC schema and validator for this.
+// The validation logic is a little bit tortured and we do not reject extraneous
+// properties, even though we should.
```

Comment:
Sounds like a good use case for `superstruct`'s `object()` 😄

Reply:
👀